CDF documentation update

bryan-harter opened the pull request:
# Change Summary
Added CDF file Documentation

## Overview
Added new documentation regarding the creation of CDF files.  This goes over the required attributes to make an ISTP compliant CDF file, and how xarray_to_cdf works.  

## New Files
cdf_guide.rst

bryan-harter committed:
CDF documentation update

bryan-harter committed:
Adding the cdf_guide.rst to the toc

bryan-harter committed:
Formatting update!

bryan-harter committed:
Merge branch 'dev' into cdf_documentation

bryan-harter committed:
Updating the docs for sections I missed

greglucas commented on docs/source/_static/2d_spectrogram_1.gif:
Can you link out to these images instead, or point people to the landing page for more description: I think a similar image is here: https://spdf.gsfc.nasa.gov/istp_guide/1d_spectrogram.GIF

greglucas commented on docs/source/_static/cdf_variables.png:
I don't think this is a very descriptive image... I find it hard to understand what this means or what the variables are referring to. Might help to link out to a different image on the CDF website for this?

bryan-harter replied:
I figured the text below it would help describe it?

greglucas replied:
I guess I just don't personally think it adds a lot to the context. I think people can grasp shape/dimensions of arrays, and if they don't I am not so sure this image helps at all with that. It isn't stacked in a nice way to show what that means. xarray's logo is a good example of something that might be a little more descriptive and informative for an image: 
![image](https://github.com/IMAP-Science-Operations-Center/imap_processing/assets/12417828/231a58a8-e505-4c94-9f9f-b1504a577f33)
 but alas, CDF likes their 1980s images :-p

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+
+##############################
+Generating CDF Files
+##############################
```
Preferably, all underlines should be the same length as the headings for consistency.

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+
+
+##############################
+Generating CDF Files
```
Generating CDF files should be a section below CDF Files Introduction I think?

Maybe make the landing heading "CDF Guide" to correspond to what you called the file.

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+******************************
+CDF Files Introduction
+******************************
+CDF files are binary files which require special libraries to create and manipulate. Goddard Space Flight Center (GSFC) maintains these CDF libraries, and is the shepherd of the file format.
```
What does CDF stand for? I don't think is mentioned anywhere here.

bryan-harter committed:
Formatting a little nicer

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+
+CDF Ecosystem
+===============
+**CDF** refers to more than just the ".cdf" files themselves, but the entire toolkit surrounding those files.  The official toolkits are available in IDL, C, Fortran and MATLAB.
```
Does CDF really refer to more than just the format itself? I'm not so sure and don't see that mentioned anywhere... Sure there are other toolkits that help with CDF files, but I think they are external to the format.

Maybe add a link to the specific CDF landing page somewhere here too: https://cdf.gsfc.nasa.gov/

bryan-harter replied:
They call the CDF files themselves "dotCDF" files in the documentation, otherwise I believe technically "CDF" is the data abstraction and software interfaces/libraries that access those files.  So I guess it depends on how you think about it.  But here is a quotes from the user guide: 

"When people first hear the term CDF they intuitively think of data formats in the traditional sense of the word (i.e., messy/convoluted storage of data on disk or tape).  CDF is more than just a format." 
(skip a few lines) 
"In addition to being a self-describing data format, CDF is also a software library."

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+===============
+**CDF** refers to more than just the ".cdf" files themselves, but the entire toolkit surrounding those files.  The official toolkits are available in IDL, C, Fortran and MATLAB.
+
+For python implementations, there is **pycdf**, which is a part of **spacepy**.  Pycdf wraps the C libraries, and also includes various bells and whistles that help make the files a litle more python friendly.
```
Add links to these packages?

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+
+For python implementations, there is **pycdf**, which is a part of **spacepy**.  Pycdf wraps the C libraries, and also includes various bells and whistles that help make the files a litle more python friendly.
+
+There is also **cdflib**, which is a pure-python implementation of the CDF libraries. **cdflib** doesn't require installing the C libraries, and also contains the functions "xarray_to_cdf" and "cdf_to_xarray" to convert between xarray Dataset objects in python and the CDF file formats.
```
implementation of the CDF specification?

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+
+The way these are actually stored in the file are somewhat complex, but it doesn't matter to the end user.  The CDF libraries enable users to easily list and extract the variables and attributes within the CDF files.
+
+In case you have several spare hours, the internal format of CDF files are described `here <https://cdaweb.gsfc.nasa.gov/pub/software/cdf/doc/cdf391/cdf39ifd.pdf>`_.
```
```suggestion
In case you have several spare hours, the internal format of CDF files are described in the `CDF specification <https://cdaweb.gsfc.nasa.gov/pub/software/cdf/doc/cdf391/cdf39ifd.pdf>`_.
```

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,677 @@
+* Name
+   * The Name of the variable
+* Size
+   * The shape of the data, i.e. the number and the size of the dimensions.  In the picture above, zVariable 1 is 5x5, and zVariable 2 is 5x2.
```
You say "Size" here. But, then you use "size of the dimensions". Maybe "length of the dimensions" instead so you don't use size in both ways.

greglucas commented on docs/source/index.rst:
```diff
@@ -40,6 +40,7 @@ The explicit code interfaces and structure are described in the :ref:`api`.
    reference/index
    development/index
    development/doc-overview
+   development/cdf_guide.rst
```
This shows up in the header of the page when it is here. I think it might make more sense to put it under the development index.rst file.

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,686 @@
+Attributes
+-----------
+
+Attributes are essentially key:value pairs inside the CDF, similar to a python dictionary object.  For example - Mission=IMAP, or Instrument=IDEX.
```
```suggestion
Attributes are essentially key=value pairs inside the CDF, similar to a python dictionary object.  For example - Mission=IMAP, or Instrument=IDEX.
```
consistency with your later examples

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,686 @@
+ISTP Compliance
+===============
+
+ISTP compliance is a standard set of attributes to use when creating CDF files.  The goal of ISTP is that others users and programs of your CDF file will know where to look for information about the data inside.  A large part of these requirements are driven by the SPDF's display tool, `CDAWeb <https://cdaweb.gsfc.nasa.gov/cdaweb/>`_.
```
Define ISTP

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,686 @@
+cdflib.xarray_to_cdf
+*****************************
+
+The Python library `cdflib <https://github.com/MAVENSDC/cdflib>`_ is a pure-python implementation of the CDF libraries.  Of particular note is the function `xarray_to_cdf <https://cdflib.readthedocs.io/en/latest/api/cdflib.xarray.xarray_to_cdf.html#cdflib.xarray.xarray_to_cdf>`_.
```
CDF specification, not libraries?

greglucas replied:
why is xarray_to_cdf of particular note? Maybe mention something about the IMAP SDC using xarray for all internal data representations.

bryan-harter committed:
Adding links enabled in the docs

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,686 @@
+
+.. note:: the netCDF file format was created by NOAA a few years after the CDF file format to plug in some of the shortfallings of the CDF file format.  These days, the orginal reasons for the split between the two formats have largely disappeared.
+
+xarray_to_cdf
```
cdflib.xarray_to_cdf ? Link to your code here as well if there is documentation for this method already?

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,686 @@
+
+* This code does not check the values *within* the attributes (except VAR_TYPE and DEPEND_{i}), only that the attributes exist!
+* This function does not put the "Epoch" variable as the first thing in the file, which was recommended (but not required) by the SPDF
+* This code creates variables in a row-major format.  Column-major is recommended (but not required) by the SPDF.
```
Since we are just getting started, could we change to what they recommend?

bryan-harter replied:
I can if I have time or if someone else wants to open a pull request on cdflib, but:

1) cdflib doesn't currently write column-major  
2) xarray_to_cdf loops through the variables in the xarray and writes them in whatever order depending on what it reads out of the dictionary first

I was somewhat surprised that those were recommended, since the software libraries abstract that stuff away.

greglucas commented on docs/source/development/cdf_guide.rst:
```diff
@@ -0,0 +1,686 @@
+================
+This attribute identifies the name of the instrument or sensor that collected the data.  Both a long name and a short name are given.  For any data file, only a single value is allowed.
+
+For IMAP, the following are valid -
```
valid, or what must be used?

greglucas reviewed (commented):
Overall I think this is a nice overview. I personally like linking out to external places where possible and only adding explanatory text/relevant text to the mission. Right now this page is huge with a lot of repeated text/images from the main sites, and I don't think people like landing on long docs pages because they get intimidated. You could break this up into a few different sub-pages / sections.

Brief CDF introduction (lots of linking out to the main SPDF / CDF pages)
cdflib / xarray section (some linking out to cdflib documentation)
IMAP standards (more text specifically relating to this mission, possibly linking to our cdf module)

bryan-harter committed:
Fixing duplicate headers

bryan-harter committed:
Getting rid of duplicate headers

bryan-harter committed:
Fixing more documentation

bryan-harter committed:
Responding to a few code review comments

bryan-harter commented:
So I think I addressed most of your comments thus far, I just need to break up the document into sub-docs.  But here is a list of things I included in the document that are things we are not currently doing (or at least I didn't do on IDEX):

- The file naming convention is now imap_{instrument}_{level}_{mode}_{description}_YYYYMMDD_{version}.cdf
- There is a variable naming convention now.  MMS had this, and it appears others thought this would be a good idea
- New global variables, one for "parents" of the file (i.e. the inputs), and one for the file's creation date.

bryan-harter commented:
Oh additionally I added the sphinx mod "sphinx.ext.autosectionlabel", so we can just link straight to the titles of sections themselves without including all those ".. _insert_title_here:" things in the rst files.  I'm happy to change it back though if there was a reason we weren't using that though.

bryan-harter committed:
Merge branch 'dev' into cdf_documentation

bryan-harter committed:
Breaking up the page into sub-pages

bryan-harter committed:
Adding a link to xarray_to_cdf page

bryan-harter committed:
Adding more links to different sections

bryan-harter committed:
fixing image link

bryan-harter commented:
Broke the big document into 4 pages:

https://imap-processing--273.org.readthedocs.build/en/273/development/CDFs/cdf_introduction.html
https://imap-processing--273.org.readthedocs.build/en/273/development/tools/xarray_to_cdf.html
https://imap-processing--273.org.readthedocs.build/en/273/development/CDFs/cdf_requirements.html
https://imap-processing--273.org.readthedocs.build/en/273/development/CDFs/cdf_verification.html

bryan-harter committed:
Editting some of the wording now that we're on different pages

greglucas commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+IMAP is archiving its science data with the `Space Physics Data Facility (SPDF) <https://spdf.gsfc.nasa.gov/>`_, which requires that data be in the CDF file format.
+
+CDF Ecosystem
+===============
```
Most of the headings are still overflowing the titles by one or two characters.

greglucas commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+.. note::
+   The L0->L2 processing will be using cdflib to generate CDF files on IMAP.  For more information about cdflib and the xarray_to_cdf function, see :ref:`cdflib.xarray_to_cdf`
+
+The other main python library is `**pycdf** <https://spacepy.github.io/pycdf.html>`_, which is a part of **spacepy**.  Pycdf wraps the C libraries, and therefore it requires downloading the C libraries.  Because it is a part of Spacepy, Pycdf requires downloading and installing Spacepy.  Pycdf includes various bells and whistles that help make the files more python friendly, and includes function ...
```
This shows up as `**pycdf**` in the built docs. I think you might be able to bold it by moving the double star outside the link target. But I don't think you even need to do that and could just let the link target choose how it wants to render since they are usually already italicized or something else.

maxinelasp replied:
```suggestion
The other main python library is `pycdf <https://spacepy.github.io/pycdf.html>`_, which is a part of **spacepy**.  Pycdf wraps the C libraries, and therefore it requires downloading the C libraries.  Because it is a part of Spacepy, Pycdf requires downloading and installing Spacepy.  Pycdf includes various bells and whistles that help make the files more python friendly, and includes functions for checking for ISTP compliance.
```

greglucas commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+
+The way these are astored in the file are somewhat complex, but that information is abstracted away from the end user.  The CDF libraries have functions to easily list and extract the variables and attributes within the CDF files.
+
+The internal format of CDF files are described `cdf specification <https://cdaweb.gsfc.nasa.gov/pub/software/cdf/doc/cdf391/cdf39ifd.pdf>`_.
```
```suggestion
The internal format of CDF files are described in the `cdf specification <https://cdaweb.gsfc.nasa.gov/pub/software/cdf/doc/cdf391/cdf39ifd.pdf>`_.
```

greglucas commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+Attributes
+-----------
+
+Attributes are essentially key:value pairs inside the CDF, similar to a python dictionary object.  For example - Mission=IMAP, or Instrument=IDEX.
```
```suggestion
Attributes are essentially (key=value) pairs inside the CDF, similar to a python dictionary object.  For example - Mission=IMAP, or Instrument=IDEX.
```
or put colons in the latter examples:
```suggestion
Attributes are essentially key:value pairs inside the CDF, similar to a python dictionary object.  For example - Mission:IMAP, or Instrument:IDEX.
```

greglucas commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+* Name
+   * Variables are given a unique name
+* Shape
+   * The the number and the size of the dimensions.  In the picture above, "zVariable 1" is 5x5, and "zVariable 2" is 5x2.
```
```suggestion
   * The number and the size of the dimensions.  In the picture above, "zVariable 1" is 5x5, and "zVariable 2" is 5x2.
```

greglucas commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+#####################
+CDF File Requirements
```
IMAP-Specific CDF Requirements ?

Do we want to be explicit that these are IMAP specific requirements since you say these are necessary for this mission in the next sentence?

greglucas commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+
+ISTP Compliant Global Attributes are listed here: `https://spdf.gsfc.nasa.gov/istp_guide/gattributes.html <https://spdf.gsfc.nasa.gov/istp_guide/gattributes.html>`_, and notes about how they are used on IMAP are below -
+
+Data_type
```
Would this be easier to follow as a table?
Something like three rows: Attribute Name, Valid Options, Description

It is always a pain to make rst tables though, so this doesn't have to happen now and can be a follow-up if someone does want to do that later.

greglucas commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+
+
+******************************
+File Naming Convention
```
I'm wondering if some of this is getting into the `style-guide` territory? I'm not sure where I would expect things to live though.

bourque replied:
We could add a link to this section the `Naming Conventions` doc in the style guide.

greglucas commented on docs/source/development/CDFs/cdf_verification.rst:
```diff
@@ -0,0 +1,31 @@
+SPDF Validation
+******************************
+
+As a final validation step, the SPDF will review all completed data produts.  They will run the CDF file through the SKTeditor as a first pass, and also ensure that the auto-generated plots look nice on CDAWeb using the IDL tool `https://cdaweb.gsfc.nasa.gov/cdfx/ <https://cdaweb.gsfc.nasa.gov/cdfx/>`_.
```
```suggestion
As a final validation step, the SPDF will review all completed data produts.  They will run the CDF file through the SKTeditor as a first pass, and also ensure that the auto-generated plots look nice on CDAWeb using the IDL tool `CDFx <https://cdaweb.gsfc.nasa.gov/cdfx/>`_.
```

bourque replied:
```suggestion
As a final validation step, the SPDF will review all completed data products.  They will run the CDF file through the SKTeditor as a first pass, and also ensure that the auto-generated plots look nice on CDAWeb using the IDL tool `https://cdaweb.gsfc.nasa.gov/cdfx/ <https://cdaweb.gsfc.nasa.gov/cdfx/>`_.
```

greglucas commented on docs/source/development/tools/xarray_to_cdf.rst:
This seems like awesome documentation for the cdflib, so I would look into moving it into there if you get the chance. Then we can cross-reference that content from IMAP. But this is more generally useful than just imap_processing IMO.

greglucas reviewed (approved):
Nice, I like the content broken up into sub-pages, much easier to digest!

A few very minor nits and comments to consider for the future, but looks good to go currently.

maxinelasp commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+===============
+**CDF** refers to more than just the files themselves, but the entire toolkit surrounding those files.  The official toolkits are available in IDL, C, Fortran and MATLAB.
+
+For python implementations, there is `**cdflib** <https://github.com/MAVENSDC/cdflib>`_, which is a pure-python implementation of the CDF specification. **cdflib** contains the functions "xarray_to_cdf" and "cdf_to_xarray" to convert between xarray Dataset objects in python and the CDF file formats.
```
```suggestion
For python implementations, there is `cdflib <https://github.com/MAVENSDC/cdflib>`_, which is a pure-python implementation of the CDF specification. **cdflib** contains the functions "xarray_to_cdf" and "cdf_to_xarray" to convert between xarray Dataset objects in python and the CDF file formats.
```

maxinelasp commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+
+The other main python library is `**pycdf** <https://spacepy.github.io/pycdf.html>`_, which is a part of **spacepy**.  Pycdf wraps the C libraries, and therefore it requires downloading the C libraries.  Because it is a part of Spacepy, Pycdf requires downloading and installing Spacepy.  Pycdf includes various bells and whistles that help make the files more python friendly, and includes function ...
+
+Because the L0->L2 processing on IMAP will be making heavy use of "xarray", it will also be using "cdflib.xarray_to_cdf" to create the ".cdf" files.
```
```suggestion

```
I think this line is unneeded and kind of confusing

maxinelasp commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+
+To archive at the SPDF, files must be ISTP compliant.  The SPDF will reject files that do not conform to their standards.
+
+A guide to SPDF can be found here: `https://spdf.gsfc.nasa.gov/istp_guide/istp_guide.html <https://spdf.gsfc.nasa.gov/istp_guide/istp_guide.html>`_
```
```suggestion
A guide to SPDF can be found `on the SPDF website. <https://spdf.gsfc.nasa.gov/istp_guide/istp_guide.html>`_
```

bourque commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+* Records
+   * Variables can span across multiple records.  This is essentially another dimension of the data, but the idea is that each record represents 1 unique physical measurement from an instrument.  In the above diagram, both variables have 3 records shown.
+
+.. note:: For IMAP instruments, each "Record" is always a unit of time.  If a varible does not vary over time, it will only span 1 record.
```
```suggestion
.. note:: For IMAP instruments, each "Record" is always a unit of time.  If a variable does not vary over time, it will only span 1 record.
```

bourque commented on docs/source/development/CDFs/cdf_introduction.rst:
```diff
@@ -0,0 +1,81 @@
+
+There are two parts of a CDF file that a user of the file ultimately cares about: Variables and Attributes.
+
+The way these are astored in the file are somewhat complex, but that information is abstracted away from the end user.  The CDF libraries have functions to easily list and extract the variables and attributes within the CDF files.
```
```suggestion
The way these are stored in the file are somewhat complex, but that information is abstracted away from the end user.  The CDF libraries have functions to easily list and extract the variables and attributes within the CDF files.
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+
+For sample code that will generate a file that matches the below guidelines, see :ref:`IMAP xarray_to_cdf Example`
+
+******************************
+Global Attributes
+******************************
```
```suggestion
*****************
Global Attributes
*****************
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+==========
+This attribute is used by CDF file writing software to create a filename. It is a combination of the following filename components: mode, data level, and optional data product descriptor.
+
+Data_version
+================
```
```suggestion
Data_version
============
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+================
+This attribute identifies the version of a particular CDF data file.
+
+Descriptor
+================
```
```suggestion
Descriptor
==========
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+* IDEX>Interstellar Dust Experiment
+* SWE>Solar Wind Electrons
+* SWAPI>Solar wind and Pickup Ions
+* CoDICE>Compact Dual Ion Compoition Experiment
```
```suggestion
* CoDICE>Compact Dual Ion Composition Experiment
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+* SWAPI>Solar wind and Pickup Ions
+* CoDICE>Compact Dual Ion Compoition Experiment
+* MAG>Magnetometer
+* HIT>High-energy Ion Teleccope
```
```suggestion
* HIT>High-energy Ion Telescope
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+* IMAP-Lo>Interstellar Mapping and Acceleration Probe Low
+* IMAP-Ultra>Interstellar Mapping and Acceleration Probe Ultra
+
+Discipline
+================
```
```suggestion
Discipline
==========
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+================
+For IMAP, this value should always be “Space Physics>Heliospheric Physics.”. This attribute describes both the science discipline and sub discipline.
+
+Generation_date
+================
```
```suggestion
Generation_date
===============
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+================
+Date stamps the creation of the file using the syntax yyyymmdd, e.g. 20150923.
+
+Instrument_type
+================
```
```suggestion
Instrument_type
===============
```

bourque replied:
Looks like there are several more of these below, so I will stop commenting on them individually 😅

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+
+MODS
+================
+This attribute is an SPDF standard global attribute, which is used to denote the history of modifications made to the CDF data set. The MODS attribute should contain a description of all significant changes to the data set, essentially capturing a log of highlevel release notes. This attribute can have as many entries as necessary and should be updated if the "X" value of the version number chang ...
```
```suggestion
This attribute is an SPDF standard global attribute, which is used to denote the history of modifications made to the CDF data set. The MODS attribute should contain a description of all significant changes to the data set, essentially capturing a log of high-level release notes. This attribute can have as many entries as necessary and should be updated if the "X" value of the version number changes.
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+================
+This attribute lists the parent data files for files of derived and merged data sets. The syntax for a CDF parent is: "CDF>logical_file_id". Multiple entry values are used for multiple parents. This attribute is required for any MMS data products that are derived from 2 or more data sources and the file names of parent data should be clearly identified. CDF parents may include source files with n ...
+
+******************************
+IMAP Variables
+******************************
```
```suggestion
***************
IMAP Variables
***************
```

bourque commented on docs/source/development/CDFs/cdf_requirements.rst:
```diff
@@ -0,0 +1,343 @@
+
+Required Epoch Variable
+------------------------
+All IMAP CDF Data files must contain at least one cariable of data type CDF_TIME_TT2000 named "Epoch".  All time varying variables in the CDF data set will depend on either this "Epoch" or another variable of type CDF_TIME_TT2000.  More than one CDF_TIME_TT2000 variable is allowed in a data set to allow for more than one time resolution.  It is recommended that all such time variable use "Epoch"  ...
```
```suggestion
All IMAP CDF Data files must contain at least one variable of data type CDF_TIME_TT2000 named "Epoch".  All time varying variables in the CDF data set will depend on either this "Epoch" or another variable of type CDF_TIME_TT2000.  More than one CDF_TIME_TT2000 variable is allowed in a data set to allow for more than one time resolution.  It is recommended that all such time variable use "Epoch" within their variable name.
```

bourque commented on docs/source/development/CDFs/cdf_verification.rst:
```diff
@@ -0,0 +1,31 @@
+The following page describes methods for verifying CDF files after they are created by IMAP processing software.
+
+
+******************************
+SKTeditor
+******************************
```
```suggestion
**********
SKTeditor
**********
```

bourque commented on docs/source/development/CDFs/cdf_verification.rst:
```diff
@@ -0,0 +1,31 @@
+You can download this tool from the link above, and open the created file there.  You can click "Show Messages" in the lower right-hand corner, and it will provide details about any ISTP compliance errors that the code may have missed.
+
+
+******************************
+SPDF Validation
+******************************
```
```suggestion
****************
SPDF Validation
****************
```

bourque commented on docs/source/development/index.rst:
```diff
@@ -41,4 +39,5 @@ A typical development workflow might look like the following:
     release-workflow
     style-guide/style-guide
     tools/index
+    CDFs/index
     poetry
```
Since we now have quite a growing list here, maybe we should just put these in alphabetical order?

bourque commented on docs/source/development/style-guide/checklist-for-pull-requests.rst:
```diff
@@ -19,10 +19,10 @@ Checklist for Contributors
 
 
-.. _Pertaining-to-the-code:
+.. _contributing-to-the-code:
```
Nice, I like this change!

bourque commented on docs/source/development/tools/xarray_to_cdf.rst:
```diff
@@ -0,0 +1,239 @@
+         "These are used by the variable imap_jim_counts_(mode)_(description)"
+         "Ideally we would actually have additional variables that describe "
+         "the minimum and the maximum values in each of these bins, but for our "
+         "purposes of creating a minimal example, we'll leave extra varaibles out. "
```
```suggestion
         "purposes of creating a minimal example, we'll leave extra variables out. "
```

bourque reviewed (changes requested):
Awesome work here @bryan-harter, this is all very detailed and helpful!  Most of my comments are just spelling and formatting related, but the contents look great.

tech3371 commented on docs/source/development/tools/xarray_to_cdf.rst:
```diff
@@ -0,0 +1,239 @@
+IMAP xarray_to_cdf Example
+***************************
+
+Suppose there is an instrument on IMAP named JIM that took 3 samples of ions at 3 different energies, such that the data looks like the following -
```
:) JIM craft(like S/C)

bryan-harter replied:
I stole the idea from Maxine, who gave Heather a test for something involving a JIM satellite

bryan-harter replied:
I stole the idea from Maxine, who gave Heather a test to design an SDC for a JIM instrument

tech3371 commented on docs/source/development/tools/xarray_to_cdf.rst:
```diff
@@ -0,0 +1,239 @@
+                           "imap_jim_counts_mode_descriptor": counts_xr,
+                     },
+                     coords={
+                           "Epoch": epoch_xr,
+                           "Energy": energy_xr
```
Does coordinates look up by 'name' and not the `dims`? The reason I ask is I noticed that the coordinate's name and `xr.DataArray`'s `dims` name did not match but it match with `xr.DataArray`'s `name`.

bryan-harter replied:
In general, the coordinate's name and the dimension's name are separate things.   You could name them whatever you wanted for our purposes, as long as you are consistent within the xarray object.  The actual dimension names ("energy_dim" in this case) doesn't show up anywhere inside the CDF file, so it's purely for the xarray Dataset to link the two variables together.  

I admit it's a bit confusing how I set it up, because I just named one dimension "Epoch", but I added a "_dim" to  "Energy".  But my goal in doing this was to show that it doesn't ultimately matter what you name the dimensions.

tech3371 replied:
Thank you for clarifying this!

tech3371 reviewed (commented):
I think others have left great reviews. Nothing to add on my end. I just had one question.

bryan-harter committed:
Typos and format fixed

bryan-harter committed:
Another format error

bourque commented:
@all-contributors please add @bryan-harter for documentation
